model/task_history.go

```diff
@@ -389,35 +391,23 @@ func (self *taskHistoryIterator) GetDistinctTestNames(ctx context.Context, env e
 			{"$unwind": fmt.Sprintf("$%v", testResultsKey)},
 			{"$group": bson.M{"_id": fmt.Sprintf("$%v.%v", testResultsKey, task.TestResultTestFileKey)}},
 		},
-		opts,
 	)
-	if err != nil {
-		return nil, errors.Wrapf(err, "error getting cursor")
-	}
-	if cursor == nil {
-		return nil, errors.New("nil cursor returned")
-	}
-	defer cursor.Close(ctx)
-
-	names := []string{}
-	for cursor.Next(ctx) {
-		select {
-		case <-ctx.Done():
-			return nil, errors.New("context cancelled")
-		default:
-			res := bson.M{}
-			if err := cursor.Decode(&res); err != nil {
-				return nil, errors.Wrapf(err, "error decoding result")
-			}
-			// remove quotes from value
-			val := strings.Replace(res["_id"].(string), "\"", "", -1)
-			names = append(names, val)
+	pipeline.MaxTime(time.Minute)
+
+	var names []string
+	res := bson.M{}
+	iter := pipeline.Iter()
+	for iter.Next(&res) {
+		names = append(names, res["_id"].(string))
+		if ctx.Err() != nil {
+			catcher := grip.NewBasicCatcher()
+			catcher.Add(errors.Wrap(ctx.Err(), "iterating results"))
+			catcher.Add(iter.Close())
```

Comment:
are we able to just defer iter.Close() ? I imagine we should do this in the non error case too

Reply:
This is an interesting question.

iter.Close() passes through to [close the cursor](https://github.com/mongodb/anser/blob/7c3b6870adfb65e15f5b130bae0adbb7efc67718/db/wrapper.go#L342). In the happy path where where the cursor gets exhausted I think the cursor gets closed automatically by the server. It was hard to find anything in the mongodb docs 😞  (I did find [this](https://docs.mongodb.com/manual/core/tailable-cursors/#:~:text=by%20default%2C%20mongodb%20will%20automatically%20close%20a%20cursor%20when%20the%20client%20has%20exhausted%20all%20results%20in%20the%20cursor)), and the internet wasn't terribly helpful, but I found [this](https://jira.mongodb.org/browse/SERVER-8968?focusedCommentId=298093&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298093) from a while ago. I determined it empirically on our db, though
```
mci:SECONDARY> var cur = db.tasks.find().limit(1000);
mci:SECONDARY> cur.isClosed()
false
mci:SECONDARY> cur.forEach((res)=>{});
mci:SECONDARY> cur.isClosed()
true
```